craft_providers/bases/buildd.py

```diff
@@ -0,0 +1,522 @@
+    :param command_environment: Additional environment to configure for command.
+        If specifying an environment, default_command_environment() is provided
+        for the minimum required environment configuration.
+    """
+
+    def __init__(
+        self,
+        *,
+        alias: BuilddBaseAlias,
+        hostname: str = "craft-buildd-instance",
+        command_environment: Optional[Dict[str, Optional[str]]] = None,
+    ):
+        self.alias: BuilddBaseAlias = alias
+        self.hostname: str = hostname
+
+        if command_environment is not None:
+            self.command_environment = command_environment
+        else:
+            self.command_environment = default_command_environment()
+
+    def setup(
+        self,
+        *,
+        executor: Executor,
+        retry_wait: float = 0.25,
+        timeout: Optional[float] = None,
+    ) -> None:
+        """Prepare base instance for use by the application.
+
+        Wait for environment to become ready and configure it.  At completion of
+        setup, the executor environment should have networking up and have all
+        of the installed dependencies required for subsequent use by the
+        application.
+
+        Setup may be called more than once in a given instance to refresh/update
+        the environment.
+
+        If timeout is specified, abort operation if time has been exceeded.
+
+        Guarantees provided by this setup:
+
+            - configured hostname
+
+            - networking available (IP & DNS resolution)
+
+            - apt cache up-to-date
+
+            - snapd configured and ready
+
+            - system services are started and ready
+
+        :param executor: Executor for target container.
+        :param retry_wait: Duration to sleep() between status checks (if
+            required).
+        :param timeout: Timeout in seconds.
+
+        :raises BaseConfigurationError: on other unexpected error.
+        """
+        if timeout is not None:
+            deadline: Optional[float] = time.time() + timeout
+        else:
+            deadline = None
+
+        self._setup_environment(executor=executor, deadline=deadline)
+        self._setup_wait_for_system_ready(
+            executor=executor, deadline=deadline, retry_wait=retry_wait
+        )
+        self._setup_hostname(executor=executor, deadline=deadline)
+        self._setup_resolved(executor=executor, deadline=deadline)
+        self._setup_networkd(executor=executor, deadline=deadline)
+        self._setup_wait_for_network(
+            executor=executor, deadline=deadline, retry_wait=retry_wait
+        )
+        self._setup_apt(executor=executor, deadline=deadline)
+        self._setup_snapd(executor=executor, deadline=deadline)
+
+    def _setup_apt(self, *, executor: Executor, deadline: Optional[float]) -> None:
+        """Configure apt & update cache.
+
+        :param executor: Executor for target container.
+        """
+        executor.create_file(
```

Comment:
Maybe we should check if deadline was met also at the start of each call so that nothing would be executed in that case?

Reply:
Fixed!

Reply:
To prevent redundant checks, I decided to call deadline checks after each command.  But it always bothered me the last operation might succeed and deadline will raise error.  I moved all checks to the front.